### PR TITLE
Out-of-order message filtering

### PR DESCRIPTION
**Pull request by thomasst**

### CODE REVIEW

File: socketshark/subscription.py [hunk 1]
```diff
@@ -78,6 +81,38 @@ def prepare_service_data(self):
         return await self.perform_service_request('on_unsubscribe',
                                                   raise_error=False)
 
+    def filter_message(self, data):
```
Comment (tsx) [inline socketshark/subscription.py]: I'd name it so something `should_drop_message` as it now returns `True` to mean "drop" message, while python's `filter` drops items when predicate evaluates to a falsy value

File: socketshark/subscription.py [hunk 2]
```diff
@@ -78,6 +81,38 @@ def prepare_service_data(self):
...
+        """
+        Returns whether to filter the given message.
+        """
+        # Check whether the message is filtered by comparing any defined
+        # filter_fields to auth_info.
+        filter_fields = self.service_config.get('filter_fields', [])
```
Comment (tsx) [inline socketshark/subscription.py]: Separate to independent functions like `drop_based_on_field` and `drop_out_of_order`.

File: socketshark/subscription.py [hunk 3]
```diff
@@ -78,6 +81,38 @@ def prepare_service_data(self):
...
+        for field in filter_fields:
+            if field in data:
+                if self.session.auth_info.get(field) != data[field]:
+                    # Message filtered.
+                    print('filtered based on', field)
```
Comment (tsx) [inline socketshark/subscription.py]: logger?

File: socketshark/subscription.py [hunk 4]
```diff
@@ -78,6 +81,38 @@ def prepare_service_data(self):
...
+            try:
+                order = int(data['_order'])
+            except (TypeError, ValueError):
+                return True  # Filter invalid orders.
+
+            if last_order is not None and order <= last_order:
+                # Message filtered.
+                return True
+
+            self.order_state[key] = order
```
Comment (tsx) [inline socketshark/subscription.py]: Is this ever cleaned up? Sounds like potential memory leak.

Comment (thomasst) [inline socketshark/subscription.py]: It should get cleaned up when the subscription is cleaned up. How do you see a memory leak? The destructor (`Subscription.__del__`) gets called in my tests when unsubscribing, and I'd assume the order dict would get cleaned up with it.

**Comment by thomasst** [comment]
@tsx check it out again!

File: socketshark/subscription.py (hunk 4 above)
Comment (tsx) [inline socketshark/subscription.py]: I was thinking it's an instance-global dictionary - you're right.

**Pull request merged**